src/offset/local/tzdata.rs

```diff
@@ -0,0 +1,131 @@
+                return Ok(file);
+            }
+        }
+    }
+    Err(Error::Io(io::Error::from(io::ErrorKind::NotFound)))
+}
+
+fn find_tz_data_in_file(file: &mut File, tz_name: &str) -> Result<Vec<u8>, Error> {
+    let header = Header::new(file)?;
+    let index = Index::new(file, header)?;
+    if let Some(entry) = index.find_entry(tz_name) {
+        file.seek(SeekFrom::Start((entry.offset + header.data_offset) as u64))?;
+        let mut tz_data = vec![0u8; entry.length];
+        file.read_exact(&mut tz_data)?;
+        Ok(tz_data)
+    } else {
+        Err(Error::Io(io::Error::from(ErrorKind::NotFound)))
+    }
+}
+
+impl Header {
+    fn new(file: &mut File) -> Result<Self, Error> {
+        let mut buf = [0; HEADER_SIZE];
+        file.read_exact(&mut buf)?;
+        if !buf.starts_with(b"tzdata") || buf[MAGIC_SIZE - 1] != 0u8 {
+            return Err(Error::InvalidTzdataFile("invalid magic number"));
+        }
+        Ok(Self {
+            index_offset: parse_tz_int(&buf, MAGIC_SIZE) as usize,
+            data_offset: parse_tz_int(&buf, MAGIC_SIZE + TZ_INT_SIZE) as usize,
+            _zonetab_offset: parse_tz_int(&buf, MAGIC_SIZE + 2 * TZ_INT_SIZE) as usize,
+        })
+    }
+}
+
+impl Index {
+    fn new(file: &mut File, header: Header) -> Result<Self, Error> {
+        file.seek(SeekFrom::Start(header.index_offset as u64))?;
+        let size = (header.data_offset - header.index_offset) as usize;
+        let mut bytes = vec![0; size];
+        file.read_exact(&mut bytes)?;
+        Ok(Self(bytes))
+    }
+
+    fn find_entry(&self, name: &str) -> Option<IndexEntry> {
+        let name_bytes = name.as_bytes();
+        let name_len = name_bytes.len();
+        if name_len > TZ_NAME_SIZE {
+            return None;
+        }
+
+        let zeros = [0u8; TZ_NAME_SIZE];
+        let cmp = |chunk: &&[u8]| -> Ordering {
+            // tz names always have TZ_NAME_SIZE bytes and are right-padded with 0s
+            // so we check that a chunk starts with `name` and the remaining bytes are 0
+            chunk[..name_len]
+                .cmp(name_bytes)
+                .then_with(|| chunk[name_len..TZ_NAME_SIZE].cmp(&zeros[name_len..]))
+        };
+
+        let chunks: Vec<_> = self.0.chunks_exact(INDEX_ENTRY_SIZE).collect();
+        chunks.binary_search_by(cmp).map(|idx| IndexEntry::new(chunks[idx])).ok()
+    }
+}
+
+impl<'a> IndexEntry<'a> {
+    fn new(bytes: &'a [u8]) -> Self {
+        Self {
+            _name: bytes[..TZ_NAME_SIZE].splitn(2, |&b| b == 0u8).next().unwrap(),
+            offset: parse_tz_int(bytes, TZ_NAME_SIZE) as usize,
+            length: parse_tz_int(bytes, TZ_NAME_SIZE + TZ_INT_SIZE) as usize,
+            _raw_utc_offset: parse_tz_int(bytes, TZ_NAME_SIZE + 2 * TZ_INT_SIZE) as usize,
+        }
+    }
+}
+
+/// Panics if slice does not contain [TZ_INT_SIZE] bytes beginning at start.
+fn parse_tz_int(slice: &[u8], start: usize) -> u32 {
```

Comment:
I'd make this a newtype with a constructor, we can use the type for the `Header` fields.